katcp/core.py

```diff
@@ -299,7 +299,23 @@ def format_argument(self, arg):
         elif isinstance(arg, bool):
             return str(int(arg))
         else:
-            return str(arg)
+            try:
+                return str(arg)
+            except UnicodeEncodeError:
+                # unicode characters will break the str cast, so
+                # try to encode to ascii and replace the offending characters
+                # with a '?' character
+                logger.exception("Error casting message argument to str!")
+                try:
+                    if not isinstance(arg, unicode):
+                        arg = arg.decode('utf-8')
+                    return arg.encode('ascii', 'replace')
+                except Exception:
+                    # If all else fails, ignore the string completely
+                    logger.exception(
+                        "Error formatting argument as str, "
+                        "ignoring message and returning ''")
+                    return ''
```

Comment:
Do we really want to continue if this happens?  Things are really broken in this case, and we need to investigate.  We have no idea what the message handler will do with the empty string argument we're passing back.

Reply:
The reason we continue is so that a sync can finish even when we get this encoding error. If we just raise a hard exception here, we get stuck in an infinite loop trying to sync with the offending device.

Reply:
The sync problem was with unicode characters, which is handled now, in the "normal" case.  If it can't handle the unicode, i.e. this last catch all exception, then something really unusual is happening, isn't it?  Not syncing may be better than continuing silently.

On the continuous re-syncing issue, what about backing off for longer before retrying if we get an exception during sync?

Reply:
I think you are right that it will be good enough to only handle the `UnicodeEncodeError` here and leave other exceptions unhandled. We definitely need to somehow spot when a device's messages are that bad. Besides, I'm not even sure how to get to the second except block.

I'm going to leave out digging into re-syncing now because I suspect that is a client issue that we need to investigate when we see it again.

Reply:
OK, we can do that.